aws/hll/aws-s3-transfer-manager/src/io/part_reader.rs

```diff
@@ -0,0 +1,336 @@
+            unsafe { dst.set_len(dst.capacity()) }
+            file_util::read_file_chunk_sync(dst.deref_mut(), path, offset)?;
+            let data = dst.freeze();
+            Ok::<PartData, Error>(PartData { data, part_number })
+        });
+
+        handle.await?.map(Some)
+    }
+}
+
+mod file_util {
+    #[cfg(unix)]
+    pub(super) use unix::read_file_chunk_sync;
+    #[cfg(windows)]
+    pub(super) use windows::read_file_chunk_sync;
+
+    #[cfg(unix)]
+    mod unix {
+        use std::fs::File;
+        use std::io;
+        use std::os::unix::fs::FileExt;
+        use std::path::Path;
+
+        pub(crate) fn read_file_chunk_sync(
+            dst: &mut [u8],
+            path: impl AsRef<Path>,
+            offset: u64,
+        ) -> Result<(), io::Error> {
+            let file = File::open(path)?;
+            file.read_exact_at(dst, offset)
+        }
+    }
+
+    #[cfg(windows)]
+    mod windows {
+        use std::fs::File;
+        use std::io;
+        use std::io::{Read, Seek, SeekFrom};
+        use std::path::Path;
+
+        pub(crate) fn read_file_chunk_sync(
+            dst: &mut [u8],
+            path: impl AsRef<Path>,
+            offset: u64,
+        ) -> Result<(), io::Error> {
+            let mut file = File::open(path)?;
+            file.seek(SeekFrom::Start(offset))?;
+            file.read_exact(dst)
+        }
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use std::io::Write;
+
+    use bytes::{Buf, Bytes};
+    use tempfile::NamedTempFile;
+
+    use crate::io::part_reader::{PartData, PartReaderBuilder, ReadPart};
+    use crate::io::InputStream;
+
+    async fn collect_parts(reader: impl ReadPart) -> Vec<PartData> {
+        let mut parts = Vec::new();
+        let mut expected_part_number = 1;
+        while let Some(part) = reader.next_part().await.unwrap() {
+            assert_eq!(expected_part_number, part.part_number);
+            expected_part_number += 1;
+            parts.push(part);
+        }
+        parts
+    }
+
+    #[tokio::test]
+    async fn test_bytes_part_reader() {
+        let data = Bytes::from("a lep is a ball, a tay is a hammer, a flix is a comb");
+        let stream = InputStream::from(data.clone());
+        let expected = data.chunks(5).collect::<Vec<_>>();
+        let reader = PartReaderBuilder::new().part_size(5).stream(stream).build();
+        let parts = collect_parts(reader).await;
+        let actual = parts.iter().map(|p| p.data.chunk()).collect::<Vec<_>>();
+
+        assert_eq!(expected, actual);
+    }
+
+    async fn path_reader_test(limit: Option<usize>, offset: Option<usize>) {
+        let part_size = 5;
+        let mut tmp = NamedTempFile::new().unwrap();
+        let mut data = Bytes::from("a lep is a ball, a tay is a hammer, a flix is a comb");
+        tmp.write_all(data.chunk()).unwrap();
+
+        let mut builder = InputStream::read_from().path(tmp.path());
+        if let Some(limit) = limit {
+            data.truncate(limit);
```

Comment:
Why do we truncate here? To simulate reading from an offset?

Reply:
never mind, I think I understand now that I've seen the invocations of this fn.

Reply:
yes it's for simulating reading up to a certain number of bytes, I'm basically abusing `Bytes` ability to (1) start reading from an offset, (2) read up to a certain amount, and (3) chunk by a certain size where all chunks except the last will be of that size (which is the exact behavior we want part readers to have).